### PR TITLE
Generalise the definition of lists of nested binders

### PR DESCRIPTION
**Pull request by AltGr**
with the addition of a new generic `BoundList` module.

### CODE REVIEW

**Comment by AltGr** [comment]
Update: added tweaks to the Makefile with a new `make alltest` rule that basically simulates the CI, fetching and testing both examples and french-law

File: compiler/plugins/explain.ml [hunk 1]
```diff
@@ -415,12 +415,12 @@ let result_level base_vars =
 let interpret_program (prg : ('dcalc, 'm) gexpr program) (scope : ScopeName.t) :
     ('t, 'm) gexpr * Env.t =
   let ctx = prg.decl_ctx in
-  let all_env, scopes =
-    Scope.fold_left prg.code_items ~init:(Env.empty, ScopeName.Map.empty)
+  let (all_env, scopes), () =
+    BoundList.fold_left prg.code_items ~init:(Env.empty, ScopeName.Map.empty)
```
Comment (denismerigoux) [inline compiler/plugins/explain.ml:419]: `BoundList.map_to_list`?

File: compiler/shared_ast/boundList.ml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+open Definitions
+
+type ('e, 'elt, 'last) t = ('e, 'elt, 'last) bound_list =
+  | Last of 'last
+  | Cons of 'elt * ('e, ('e, 'elt, 'last) t) binder
```
Comment (denismerigoux) [inline compiler/shared_ast/boundList.ml:21]: Nice

File: compiler/shared_ast/boundList.mli [hunk 3]
```diff
@@ -0,0 +1,92 @@
+(* This file is part of the Catala compiler, a specification language for tax
+   and social benefits computation rules. Copyright (C) 2020-2022 Inria,
```
Comment (denismerigoux) [inline compiler/shared_ast/boundList.mli:2]: 2020-2024

Comment (denismerigoux) [reply compiler/shared_ast/boundList.mli:2]: We should bump all header files anyway.

File: compiler/shared_ast/optimizations.ml [hunk 4]
```diff
@@ -146,7 +143,7 @@ let rec optimize_expr :
                   Some
                     (Expr.unbox
                        (Expr.make_abs [| v1 |]
-                          (Expr.box
+                          (Expr.rebox
```
Comment (denismerigoux) [inline compiler/shared_ast/optimizations.ml:146]: Nice!

File: compiler/verification/conditions.ml [hunk 5]
```diff
@@ -378,49 +376,53 @@ let generate_verification_conditions_code_items
...
+                  current_scope_name = name;
+                  decl = decl_ctx;
+                  input_vars = [];
+                  scope_variables_typs =
+                    Var.Map.empty
+                    (* We don't need to add the typ of the scope input var here
+                       because it will never appear in an expression for which
+                       we generate a verification conditions (the big struct is
+                       destructured with a series of let bindings just after.
+                       )*);
```
Comment (denismerigoux) [inline compiler/verification/conditions.ml:407]: extra newline should be removed

File: dune [hunk 6]
```diff
@@ -2,6 +2,8 @@
 
 (data_only_dirs tests syntax_highlighting)
 
+(vendored_dirs _catala-examples _french-law)
```
Comment (denismerigoux) [inline dune:5]: Fancy `dune`ing

**Review by denismerigoux — APPROVED** [review]
Nice as always, @AltGr!

File: build_system/clerk_runtest.ml [hunk 7]
```diff
@@ -94,8 +94,7 @@ let run_inline_tests catala_exe catala_opts filename =
         skip_block lines
       | Some args ->
         let args = String.split_on_char ' ' args in
-        run_catala_test catala_exe catala_opts filename
-          lines_until_now args oc;
```
[the remark this answers is not in the document]
Comment (AltGr) [reply build_system/clerk_runtest.ml:98]: See f767285 ; I think that was a lack of reformatting from the previous patch (and it seems that the new CI doesn't detect formatting diffs correctly, I'll look into it)

File: Makefile [hunk 8]
```diff
@@ -238,6 +238,27 @@ all: \
...
 	plugins
 
+#> alltest					: Runs more extensive tests, including the examples and french-law. Use before push!
+alltest:
+	@export DUNE_PROFILE=check && \
+	dune build @for-tests @install @runtest && \
+	clerk test tests && \
+	rm -rf _catala-examples _french-law && \
+	: trap "rm -rf _catala-examples _french-law $$TMP" EXIT && \
+	{ git clone https://github.com/CatalaLang/catala-examples \
```
[the remark this answers is not in the document]
Comment (AltGr) [reply Makefile:248]: hm it's slightly different in that here we'll build inline using `dune` handling of subprojects in subdirectories, while the CI will go through installations to the prefix and normal dependency handling ; that allows for much quicker `make alltest` that doesn't pollute the installation dir, while still testing basically the same artefacts.

But having the CI still test the more basic workflow with installations and completely separate projects is also worthwhile.

Another reason is that here I lookup the local git branch, but we need a specific way for doing that properly when called from Github.

File: compiler/lcalc/closure_conversion.ml [hunk 9]
```diff
@@ -298,28 +297,26 @@ let transform_closures_scope_let ctx scope_body_expr =
...
-      Bindlib.box_apply
-        (fun res -> Result res)
-        (Expr.Box.lift new_scope_let_expr))
+      Bindlib.box_apply (fun e -> Last e) (Expr.Box.lift new_scope_let_expr))
     scope_body_expr
 
 let transform_closures_program (p : 'm program) : 'm program Bindlib.box =
-  let _, new_code_items =
-    Scope.fold_map
+  let (), new_code_items =
```
[the remark this answers is not in the document]
Comment (AltGr) [reply compiler/lcalc/closure_conversion.ml:318]: The `map` function didn't give access to the variable names, but I've been fixing that already, an update is coming ^^

**Pull request merged**